src/Catalog/Helpers/GalleryDatabaseQueryService.cs

```diff
@@ -0,0 +1,166 @@
+
+            return OrderPackagesByKeyDate(allPackages, keyDateFunc);
+        }
+
+        /// <summary>
+        /// Returns a <see cref="SortedList{DateTime, IList{FeedPackageDetails}}"/> of packages.
+        /// </summary>
+        /// <param name="keyDateFunc">The <see cref="DateTime"/> field to sort the <see cref="FeedPackageDetails"/> on.</param>
+        internal static SortedList<DateTime, IList<FeedPackageDetails>> OrderPackagesByKeyDate(
+            IReadOnlyCollection<FeedPackageDetails> packages,
+            Func<FeedPackageDetails, DateTime> keyDateFunc)
+        {
+            var result = new SortedList<DateTime, IList<FeedPackageDetails>>();
+
+            foreach (var package in packages)
+            {
+                var packageKeyDate = keyDateFunc(package);
+                if (!result.TryGetValue(packageKeyDate, out IList<FeedPackageDetails> packagesWithSameKeyDate))
+                {
+                    packagesWithSameKeyDate = new List<FeedPackageDetails>();
+                    result.Add(packageKeyDate, packagesWithSameKeyDate);
+                }
+
+                packagesWithSameKeyDate.Add(package);
+            }
+
+            var packagesCount = 0;
+            var filteredResult = new SortedList<DateTime, IList<FeedPackageDetails>>();
+            foreach (var keyDate in result.Keys)
+            {
+                if (result.TryGetValue(keyDate, out IList<FeedPackageDetails> packagesForKeyDate))
+                {
+                    if (packagesCount > 0 && packagesCount + packagesForKeyDate.Count > Constants.MaxPageSize)
+                    {
+                        break;
+                    }
+
+                    packagesCount += packagesForKeyDate.Count;
+                    filteredResult.Add(keyDate, packagesForKeyDate);
+                }
+            }
+
+            return filteredResult;
+        }
+
+        /// <summary>
+        /// Builds the SQL query string for db2catalog.
+        /// </summary>
+        /// <param name="cursor">The <see cref="DB2CatalogCursor"/> to be used.</param>
+        /// <returns>The SQL query string for the db2catalog job, build from the the provided <see cref="DB2CatalogCursor"/>.</returns>
+        internal static string BuildDB2CatalogSqlQuery(DB2CatalogCursor cursor)
+        {
+            return $@"SELECT TOP {cursor.Top} WITH TIES
+                        PR.[Id],
+                        P.[NormalizedVersion],
+                        P.[Created],
+                        P.[LastEdited],
+                        P.[Published],
+                        P.[Listed],
+                        P.[HideLicenseReport],
+                        P.[LicenseNames],
+                        P.[LicenseReportUrl]
+                    FROM [dbo].[Packages] AS P
+                    INNER JOIN [dbo].[PackageRegistrations] AS PR ON P.[PackageRegistrationKey] = PR.[Key]
+                    WHERE P.[PackageStatusKey] = {PackageStatus.Available} 
```

Comment:
Does `{PackageStatus.Available}` become `0` or `Available`?

Reply:
OMG this would 've been horrible! Great catch! Fixed in https://github.com/NuGet/NuGet.Services.Metadata/pull/524/commits/6e5c213ee5e0a0cd44229a5f964c16baaf7be007